src/librustc_mir/interpret/intrinsics/type_name.rs

```diff
@@ -0,0 +1,217 @@
+use rustc::ty::{
+    TyCtxt, Ty,
+    subst::{UnpackedKind, Kind},
+    print::{Printer, PrettyPrinter, Print},
+    self,
+};
+use rustc::hir::map::{DefPathData, DisambiguatedDefPathData};
+use rustc::hir::def_id::CrateNum;
+use std::fmt::Write;
+use rustc::mir::interpret::{Allocation, ConstValue, Pointer};
+use std::convert::TryInto;
+
+struct AbsolutePathPrinter<'a, 'tcx> {
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    path: String,
+}
+
+impl<'tcx> Printer<'tcx, 'tcx> for AbsolutePathPrinter<'_, 'tcx> {
+    type Error = std::fmt::Error;
+
+    type Path = Self;
+    type Region = Self;
+    type Type = Self;
+    type DynExistential = Self;
+
+    fn tcx<'a>(&'a self) -> TyCtxt<'a, 'tcx, 'tcx> {
+        self.tcx
+    }
+
+    fn print_region(self, _region: ty::Region<'_>) -> Result<Self::Region, Self::Error> {
+        Ok(self)
+    }
+
+    fn print_type(self, ty: Ty<'tcx>) -> Result<Self::Type, Self::Error> {
+        match ty.sty {
+            // types without identity
+            | ty::Bool
+            | ty::Char
+            | ty::Int(_)
+            | ty::Uint(_)
+            | ty::Float(_)
+            | ty::Str
+            | ty::Array(_, _)
+            | ty::Slice(_)
+            | ty::RawPtr(_)
+            | ty::Ref(_, _, _)
+            | ty::FnPtr(_)
+            | ty::Never
+            | ty::Tuple(_)
+            | ty::Dynamic(_, _)
+            // should be unreachable, but there's no hurt in printing it (and better than ICEing)
+            | ty::Error
+            => self.pretty_print_type(ty),
+            | ty::Infer(_)
+            | ty::Bound(_, _)
+            | ty::Param(_)
+            | ty::Placeholder(_)
+            | ty::Projection(_)
+            | ty::UnnormalizedProjection(_)
+            | ty::GeneratorWitness(_)
+            => bug!(
+                "{:#?} in `type_name` should not happen because we are always monomorphized",
+                ty,
+            ),
+            // types with identity (print the module path instead)
+            ty::Adt(adt, substs) => self.print_value_path(adt.did, substs),
+            ty::Foreign(did) => self.print_value_path(did, &[]),
+            | ty::FnDef(did, substs)
+            | ty::Opaque(did, substs)
+            => self.print_value_path(did, substs),
+            ty::Closure(did, substs) => self.print_value_path(did, substs.substs),
+            ty::Generator(did, substs, _) => self.print_value_path(did, substs.substs),
```

Comment:
I'm a bit surprised all of these are using `print_value_path`, I would've expected perhaps only `FnDef` to?

Reply:
[You said](https://github.com/rust-lang/rust/pull/60166#discussion_r279829800) "everything with a `DefId`" :D

Generators and Closures are now platform independent (they printed their file and line info before)

Adt, Foreign, FnDef and Opaque are probably fine via the `pretty_print_type` path.